stacks/api/bucket/utilities.ts

```diff
@@ -0,0 +1,68 @@
+export function findRelations(schema: any, bucketId: string, path: string = "", targets: string[]) {
+  path = path ? `${path}.` : ``;
+  for (const key of Object.keys(schema)) {
+    if (isObject(schema[key])) {
+      findRelations(schema[key].properties, bucketId, `${path}${key}`, targets);
+    } else if (isRelation(schema[key], bucketId)) {
+      targets.push(`${path}${key}`);
+    }
+  }
+  return targets;
+}
+
+export function findRemovedKeys(
+  previousSchema: any,
+  currentSchema: any,
+  removedKeys: string[],
+  path: string
+) {
+  for (const key of Object.keys(previousSchema)) {
+    if (!currentSchema.hasOwnProperty(key)) {
+      removedKeys.push(path ? `${path}.${key}` : key);
+      //we dont need to check child keys of this key anymore
+      continue;
+    }
+    if (isObject(previousSchema[key]) && isObject(currentSchema[key])) {
+      findRemovedKeys(
+        previousSchema[key].properties,
+        currentSchema[key].properties,
+        removedKeys,
+        path ? `${path}.${key}` : key
+      );
+    } else if (isArray(previousSchema[key]) && isArray(currentSchema[key])) {
+      addArrayPattern(
+        previousSchema[key].items,
+        currentSchema[key].items,
+        removedKeys,
+        path ? `${path}.${key}` : key
+      );
+    }
+  }
+  return removedKeys;
+}
+
+export function addArrayPattern(
+  previousSchema: any,
+  currentSchema: any,
+  removedKeys: string[],
+  path: string
+) {
+  path = `${path}.$[]`;
+  if (isArray(previousSchema) && isArray(currentSchema)) {
+    addArrayPattern(previousSchema.items, currentSchema.items, removedKeys, path);
+  } else if (isObject(previousSchema) && isObject(currentSchema)) {
+    findRemovedKeys(previousSchema.properties, currentSchema.properties, removedKeys, path);
+  }
+}
+
+export function isObject(schema: any) {
+  return schema.type == "object" && Object.keys(schema.properties).length > 0 ? true : false;
```

Comment:
```suggestion
  return schema.type == "object" ;
```

Reply:
When the schema like `{ type:"object",properties:{} }`, it will be failed. I think this schema is valid but doesn't need to be checked for child keys.

Reply:
Why would it fail if there are no properties?  Also, that is can't be a case since the user has to have at least one property. so it should be safe to remove that condition.

Reply:
Because if it's an object, it will pass properties of object to this recursive method and it will be failed at `for(const key of Object.keys(schema))`. Also, the user can send this kind of schema inside of object field. For example;
1- Create a new bucket and add an object field that doesn't have any property.